WalletWasabi/Services/WalletService.cs

```diff
@@ -966,300 +901,230 @@ public Operation(Script script, Money amount, string label)
+				changeHdPubKey = KeyManager.GetKeys(KeyState.Clean, true).RandomElement();
 
-			var toRemoveList = new List<(Script script, Money money, string label)>(realToSend);
-			toRemoveList.RemoveAll(x => x.money == Money.Zero);
-			realToSend = toRemoveList.ToArray();
+				builder.SetChange(changeHdPubKey.P2wpkhScript);
+			}
 
-			// 1. Get the possible changes.
-			Script changeScriptPubKey;
-			var sb = new StringBuilder();
-			foreach (var item in realToSend)
+			builder.SendEstimatedFees(feeRate);
+
+			var psbt = builder.BuildPSBT(false);
+
+			var spentCoins = psbt.Inputs.Select(txin => smartCoinsByOutpoint[txin.PrevOut]).ToArray();
+
+			var realToSend = payments.Requests.Select(t => (label: t.Label,
+												destination: t.Destination,
+												amount: psbt.Outputs.FirstOrDefault(o => o.ScriptPubKey == t.Destination.ScriptPubKey)?.Value))
+									.Where(i => i.amount != null)
+									.ToArray();
+
+			if (!psbt.TryGetFee(out var fee))
 			{
-				var corrected = Guard.Correct(item.label);
-				sb.Append($"{corrected}, ");
+				throw new InvalidOperationException($"Impossible to get the fees of the PSBT, this should never happen.");
 			}
-			var changeLabel = sb.ToString().TrimEnd(',', ' ');
+			Logger.LogInfo<WalletService>($"Fee: {fee.Satoshi} Satoshi.");
 
-			if (customChange is null)
+			var vSize = builder.EstimateSize(psbt.GetOriginalTransaction(), true);
+			Logger.LogInfo<WalletService>($"Estimated tx size: {vSize} vbytes.");
+
+			// Do some checks
+			Money totalSendAmountNoFee = realToSend.Sum(x => x.amount);
+			if (totalSendAmountNoFee == Money.Zero)
 			{
-				KeyManager.AssertCleanKeysIndexed(isInternal: true);
-				KeyManager.AssertLockedInternalKeysIndexed(14);
-				var changeHdPubKey = KeyManager.GetKeys(KeyState.Clean, true).RandomElement();
+				throw new InvalidOperationException($"The amount after subtracting the fee is too small to be sent.");
+			}
+			Money totalSendAmount = totalSendAmountNoFee + fee;
 
-				changeHdPubKey.SetLabel(changeLabel, KeyManager);
-				changeScriptPubKey = changeHdPubKey.P2wpkhScript;
+			Money totalOutgoingAmountNoFee;
+			if (changeHdPubKey is null)
+			{
+				totalOutgoingAmountNoFee = totalSendAmountNoFee;
 			}
 			else
 			{
-				changeScriptPubKey = customChange;
+				totalOutgoingAmountNoFee = realToSend.Where(x => !changeHdPubKey.ContainsScript(x.destination.ScriptPubKey)).Sum(x => x.amount);
 			}
-
-			// 6. Do some checks
-			Money totalOutgoingAmountNoFee = realToSend.Select(x => x.amount).Sum();
-			Money totalOutgoingAmount = totalOutgoingAmountNoFee + fee;
-			decimal feePc = (100 * fee.ToDecimal(MoneyUnit.BTC)) / totalOutgoingAmountNoFee.ToDecimal(MoneyUnit.BTC);
+			decimal totalOutgoingAmountNoFeeDecimal = totalOutgoingAmountNoFee.ToDecimal(MoneyUnit.BTC);
+			decimal feePc = (100 * fee.ToDecimal(MoneyUnit.BTC)) / (totalOutgoingAmountNoFeeDecimal == 0 ? decimal.MinValue : totalOutgoingAmountNoFeeDecimal);
```

Comment:
Maybe `? decimal.MinValue` and `: totalOutgoingAmountNoFeeDecimal);` should be on two different lines for better readability.

Reply:
Done